dissect/hypervisor/disk/hdd.py

```diff
@@ -0,0 +1,404 @@
+    guid: UUID
+    parent: UUID
+
+    @classmethod
+    def from_xml(cls, element: Element) -> Shot:
+        if element.tag != "Shot":
+            raise ValueError("Invalid Shot XML element")
+
+        return cls(
+            UUID(element.find("GUID").text),
+            UUID(element.find("ParentGUID").text),
+        )
+
+
+class StorageStream(AlignedStream):
+    """Stream implementation for HDD streams.
+
+    HDD files can exist of one or multiple streams, starting at consecutive offsets.
+    This class stitches all streams together into a single stream.
+
+    Args:
+        streams: A list of :class:`Storage` and file-like object tuples.
+    """
+
+    def __init__(self, streams: list[tuple[Storage, BinaryIO]]):
+        self.streams = sorted(streams, key=lambda entry: entry[0].start)
+        self._lookup = []
+
+        size = 0
+        for storage, _ in self.streams:
+            self._lookup.append(storage.start)
+            size = storage.end
+
+        super().__init__(size * SECTOR_SIZE)
+
+    def _read(self, offset: int, length: int) -> bytes:
+        sector = offset // SECTOR_SIZE
+        count = (length + SECTOR_SIZE - 1) // SECTOR_SIZE
+
+        result = []
+        stream_idx = bisect_right(self._lookup, sector) - 1
+
+        while count > 0 and stream_idx < len(self.streams):
+            storage, stream = self.streams[stream_idx]
+            sectors_remaining = storage.end - sector
+            read_sectors = min(sectors_remaining, count)
+
+            stream.seek((sector - storage.start) * SECTOR_SIZE)
+            result.append(stream.read(read_sectors * SECTOR_SIZE))
+
+            sector += read_sectors
+            count -= read_sectors
+            stream_idx += 1
+
+        return b"".join(result)
+
+
+class HDS(AlignedStream):
+    """Parallels HDS implementation.
+
+    HDS is the format for Parallels sparse disk files.
+
+    Args:
+        fh: The file-like object to the HDS file.
+        parent: Optional file-like object for the parent HDS file.
+    """
+
+    def __init__(self, fh: BinaryIO, parent: Optional[BinaryIO] = None):
+        self.fh = fh
+        self.parent = parent
+
+        self.header = c_hdd.pvd_header(fh)
+        if self.header.m_Sig not in (c_hdd.SIGNATURE_STRUCTURED_DISK_V1, c_hdd.SIGNATURE_STRUCTURED_DISK_V2):
+            raise InvalidHeaderError(f"Invalid HDS header signature: {self.header.m_Sig}")
+
+        if self.header.m_Sig == c_hdd.SIGNATURE_STRUCTURED_DISK_V1:
+            size = self.header.m_SizeInSectors_v1
+            self._bat_step = self.header.m_Sectors
+            self._bat_multiplier = 1
+        else:
+            size = self.header.m_SizeInSectors_v2
+            self._bat_step = 1
+            self._bat_multiplier = self.header.m_Sectors
+
+        self.cluster_size = self.header.m_Sectors * SECTOR_SIZE
+
+        self.data_offset = self.header.m_FirstBlockOffset
+        self.in_use = self.header.m_DiskInUse == c_hdd.SIGNATURE_DISK_IN_USE
```

Comment:
I would recommend to use braces here to visually distinguish the comparison from the assignment. But this is personal taste.

Reply:
Black removes those unfortunately, and I don't think it warrants a `# nofmt`.